app/src/main/java/com/github/ashutoshgngwr/noice/sound/PlaybackManager.kt

```diff
@@ -0,0 +1,265 @@
+      AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
+        Log.d(TAG, "Temporarily lost audio focus! Pause playback...")
+        hasAudioFocus = false
+        resumeOnFocusGain = true
+        playbackDelayed = false
+        pauseAll()
+      }
+    }
+  }
+
+  // creates audio focus request and handles its response
+  private fun requestAudioFocus() {
+    if (hasAudioFocus) {
+      return
+    }
+
+    val result = AudioManagerCompat.requestAudioFocus(audioManager, audioFocusRequest)
+    Log.d(TAG, "AudioFocusRequest result: $result")
+    when (result) {
+      AudioManager.AUDIOFOCUS_REQUEST_DELAYED -> {
+        Log.d(TAG, "Audio focus request was delayed! Pause playback for now.")
+        playbackDelayed = true
+        hasAudioFocus = false
+        resumeOnFocusGain = false
+        pauseAll()
+      }
+      AudioManager.AUDIOFOCUS_REQUEST_FAILED -> {
+        Log.d(TAG, "Failed to get audio focus! Stop playback...")
+        hasAudioFocus = false
+        playbackDelayed = false
+        resumeOnFocusGain = false
+        pauseAll()
+      }
+      AudioManager.AUDIOFOCUS_REQUEST_GRANTED -> {
+        hasAudioFocus = true
+        playbackDelayed = false
+        resumeOnFocusGain = false
+        resumeAll()
+      }
+    }
+  }
+
+  // publishes update events for clients.
+  private fun notifyChanges() {
+    eventBus.postSticky(playbacks)
+    eventBus.post(UpdateEvent(getState()))
+  }
+
+  // initializes Playback for a sound
+  private fun createPlayback(sound: Sound): Playback {
+    return Playback(context, sound, audioSessionId, audioAttributes)
+  }
+
+  // starts playing a sound. It also creates an audio focus request if we don't have it.
+  // Playback won't start immediately if audio focus is not present. We always ensure that we
+  // have audio focus before starting the playback.
+  private fun play(sound: Sound) {
+    if (!playbacks.containsKey(sound.key)) {
+      playbacks[sound.key] = createPlayback(sound)
+    }
+
+    // notify updates before any returns happen
+    notifyChanges()
+    if (playbackDelayed || resumeOnFocusGain) {
+      // If playback is paused, add this sound to playbacks and it will be played whenever the
+      // playback is resumed.
+      return
+    }
+
+    if (!hasAudioFocus) {
+      requestAudioFocus()
+      return
+    }
+
+    requireNotNull(playbacks[sound.key]).play()
+  }
+
+  // stops a playback and releases its resources.
+  private fun stop(sound: Sound) {
+    requireNotNull(playbacks[sound.key]).apply {
+      stop()
+      release()
+    }
+
+    playbacks.remove(sound.key)
+    notifyChanges()
+  }
+
+  // stops all playbacks without removing their states
+  private fun pauseAll() {
+    isPaused = true
+    playbacks.values.forEach {
+      it.stop()
+    }
+
+    notifyChanges()
+  }
+
+  // resumes all playbacks from saved state
+  private fun resumeAll() {
+    isPaused = false
+    playbacks.values.forEach {
+      it.play()
+    }
+
+    notifyChanges()
+  }
```

Comment:
There's no guarantee that we have audio  focus here. It should call `this.play(it.soundKey)` for each playback.